compiler/passes/normalize.cpp

```diff
@@ -1066,6 +1041,51 @@ static TypeSymbol* resolveTypeAlias(SymExpr* se)
   return NULL;
 }
 
+//
+// These helpers handle RiSpec (Reduce Intent Specification) i.e.:
+//
+//   forall ... with (<RiSpec> reduce <outer variable>) { ... }
+//
+// In particular, they implement RiSpecs of the form type(someArg).
+// See e.g. test/parallel/forall/vass/3types-*.
+// We want to keep these reduce intents in their original form
+// until we process reduce intents later.
+//
+// We do it here to avoid transforming it into _type_construct_C ( ... ).
+// That would be incorrect because this is a special syntax for reduce intent.
+//
+static SymExpr* callUsedInRiSpec(Expr* call, CallExpr* parent) {
+  if (parent && parent->isPrimitive(PRIM_MOVE)) {
+    SymExpr* destSE = toSymExpr(parent->get(1));
+    Symbol* dest = destSE->symbol();
+    SymExpr* riSpecMaybe = dest->firstSymExpr();
+    if (ForallStmt* fs = toForallStmt(riSpecMaybe->parentExpr)) {
+      for_vector(Expr, riSpec, fs->withClause()->riSpecs)
+        if (riSpecMaybe == riSpec)
+          return riSpecMaybe;
+    }
+  }
+  return NULL;
+}
+//
+// This function partially un-does the work of call_constructor_for_class
+// so that reduce intents specs (see above) don't get messed up.
```

Comment:
Would it be simpler to make call_constructor_for_class not transform reduce intents, instead of transforming them and then undoing it? Is there some reason that would be harder to implement?